### PR TITLE
[Not ready for Review][GPU] Sdpa causal fix

### PR DESCRIPTION
**Pull request by Kotomi-Du**
### Details:
Current SDPA micro kernel has bug on non-PageAttention + is_causal path which is observed when enabling MSFT LLM model. 

To fix the issue, couples of changes are made across different components in the PR. The main reason is because [v13::ScaledDotProductAttention](https://docs.openvino.ai/2024/documentation/openvino-ir-format/operation-sets/operation-specs/sequence/scaled-dot-product-attention.html) operator's causal mask specification does not define behavior for the LLM decoding scenario where `seq_q < seq_kv` (new tokens attending to past KV cache). Therefore, it is expected that GPU SDPA kernel doesn't cover this scenario.

However, this problem has to be addressed in general to allow LLM decoding scenario achieving best performance by using optimized causal kernel which is actually existed in CPU native SDPA kernel. More investigation can be seen in [SDPA_Causal_maks_analysis](https://github.com/Kotomi-Du/OpenVINO_Notes/blob/main/sdpa_causal_mask_analysis.md).

Currently, the PR covers the proposed solution A in readme by automatically using lower-right alignment. However this deviates from PyTorch's upper-left spec. Open for discussion to finalize the solution.

### Tickets:
 - CVS-187874 ; CVS-191886

### AI Assistance:
 - *AI assistance used: yes*
 -  AI helps create the unit test, and it is reviewed and modified by engineer.

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_micro.cl [hunk 1]
```diff
@@ -732,6 +732,9 @@ KERNEL(micro_sdpa)(OPTIONAL_SHAPE_INFO_ARG
             col_offset += k - q;
             causal_q_begin += k - q;
         #endif
+    #elif !IS_PAGED_ATTENTION && !SLIDING_WINDOW_SIZE
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_micro.cl]: minor) Do we need to check `!SLIDING_WINDOW_SIZE`? Actually, there is no such check for `causal_k` at L232.
Currently, non-PA path does not support the sliding window feature, so it's not a problem.

Comment (Kotomi-Du) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_micro.cl]: if sliding windows size can only to be true when PA is enabled, then we can remove the condition.  Code is updated.

**Comment by Kotomi-Du** [comment]
The causal path has fundamental issue in core component; switch to draft till all failure cases are fixed

**Comment by Kotomi-Du** [comment]
build_jenkins

File: tests/layer_tests/pytorch_tests/test_scaled_dot_product_attention.py [hunk 2]
```diff
@@ -59,6 +59,9 @@ def test_scaled_dot_product_atten(self, ie_device, precision, ir_version,
                                       mask, is_causal, mask_shape, dyn_shapes, enable_gqa):
         if PytorchLayerTest.use_torch_export() and not mask and is_causal:
             pytest.xfail(reason="Unsupported case for torch.export")
+        if not mask and is_causal:
+            pytest.xfail(reason="CVS-XXXXX: OpenVINO uses lower-right causal alignment "
+                         "for decoding support, differs from PyTorch upper-left when seq_q != seq_kv")
```
Comment (mitruska) [inline tests/layer_tests/pytorch_tests/test_scaled_dot_product_attention.py]: All of the `is_causal` tests have been xfailed, what means the tests shapes simulate decoding stage `seq_q < seq_kv`.
My sugestion is to confirm that prefill stage computation is still aligned by extending the test cases with prefill like shapes `seq_q == seq_kv` and avoid skipping all tests for `is_causal`.

Current shapes:
```
class TestScaledDotProductAttention(PytorchLayerTest):
    def _prepare_input(self, dtype):
        return (self.random.randn(1, 2, 8, 4, dtype=dtype),  // to be self.random.randn(1, 2, 16, 4,)
                self.random.randn(1, 2, 16, 4, dtype=dtype),
                self.random.randn(1, 2, 16, 6, dtype=dtype))
```

**Review by mitruska — COMMENTED** [review]
I understand the motivation of this change, and partially it is already applied in plugins, the only concern is to ensure backward compatibility.
It seems reasonable to me handle presented decoding case under the hood without explicit mask input (specific for decoding), for example PagedAttention doesn't even have mask input and it works as expected.
